### PR TITLE
Fixes #24886 - fixed audit in seeds test

### PR DESCRIPTION
**Pull request by lzap**
For some reason, audit was disabled randomly for all model classes. This
has to do with how Audited gem is loaded by bundler perhaps, not sure.
This patch enables it explicitly.

It also makes seeds test faster by only seeding minimum files required
for individual tests.

### CODE REVIEW

File: db/seeds.rb [hunk 1]
```diff
@@ -13,7 +13,7 @@ def format_errors(model = nil)
 end
 
 # now we load all seed files
-foreman_seeds = Dir.glob(Rails.root + 'db/seeds.d/*.rb')
+foreman_seeds = Dir.glob(Rails.root + 'db/seeds.d/*.rb').sort
```
Comment (lzap) [inline db/seeds.rb]: Oh this was left over. Removing.

File: test/unit/tasks/seeds_test.rb [hunk 2]
```diff
@@ -103,27 +110,32 @@ def seed
...
-    tmpl.name = 'test'
-    tmpl.save!
-    seed
-    refute ProvisioningTemplate.unscoped.find_by_name('Kickstart default')
+    seed('020-provisioning_templates_list.rb', '070-provisioning_templates.rb')
+    with_auditing(ProvisioningTemplate) do
+      tmpl = ProvisioningTemplate.unscoped.find_by_name('Kickstart default')
+      tmpl.name = 'test'
+      tmpl.save!
+    end
```
Comment (lzap) [inline test/unit/tasks/seeds_test.rb:132]: This is the key thing that fixes random failures - ensures that auditing is enabled during this test.

File: test/test_helper.rb [hunk 3]
```diff
@@ -160,3 +160,11 @@ def restore_plugins
...
   @klass.instance_variable_set('@registered_plugins', @plugins_backup)
 end
+
+def with_auditing(klass)
+  auditing_was_enabled = klass.auditing_enabled
+  klass.enable_auditing
+  yield
+ensure
+  klass.enable_auditing if auditing_was_enabled
+end
```
Comment (lzap) [inline test/test_helper.rb:170]: I don't think this is useful outside of testing context, so not putting this into model classes.

File: test/unit/tasks/seeds_test.rb [hunk 4]
```diff
@@ -14,10 +15,16 @@ class SeedsTest < ActiveSupport::TestCase
...
     # Authorisation is disabled usually when run from a rake db:* task
     as_admin do
-      load File.expand_path('../../../db/seeds.rb', __dir__)
+      if seed_files.empty?
+        load File.expand_path('../../../db/seeds.rb', __dir__)
+      else
+        seed_files.each do |file|
+          load File.expand_path("../../../db/seeds.d/#{file}", __dir__)
+        end
+      end
```
Comment (bastilian) [inline test/unit/tasks/seeds_test.rb:24]: This could also be written like this, saving a few lines and DRY it up.
```ruby
      seed_files = ['seeds.rb'] if seed_files.empty?
      seed_files.each do |file|
        load File.expand_path("../../../db/seeds.d/#{file}", __dir__)
      end
```

**Comment by mmoll** [comment]
tests fail reliably now :)

File: test/test_helper.rb [hunk 5]
```diff
@@ -160,3 +160,11 @@ def restore_plugins
   @klass.clear
   @klass.instance_variable_set('@registered_plugins', @plugins_backup)
 end
+
+def with_auditing(klass)
+  auditing_was_enabled = klass.auditing_enabled
+  klass.enable_auditing
+  yield
+ensure
+  klass.enable_auditing if auditing_was_enabled
```
Comment (tbrisker) [inline test/test_helper.rb]: shouldn't this be `klass.auditing_enabled = auditing_was_enabled` to ensure it is reverted to the same state as before?

File: test/unit/tasks/seeds_test.rb [hunk 6]
```diff
@@ -35,11 +39,11 @@ def seed
     seed
 
     tables.each do |model|
-      refute model.unscoped.count.zero?
+      assert_not model.unscoped.count.zero?
```
Comment (tbrisker) [inline test/unit/tasks/seeds_test.rb]: iiuc, the difference between `refute` and `assert_not` is that refute expects `false` while `assert_not` also accepts `nil`. Is there a condition in which this and other assertions you changed return `nil` instead of `false`?

File: test/unit/tasks/seeds_test.rb [hunk 7]
```diff
@@ -14,10 +15,13 @@ class SeedsTest < ActiveSupport::TestCase
     Foreman.stubs(:in_rake?).returns(true)
   end
 
-  def seed
+  def seed(*seed_files)
```
Comment (tbrisker) [inline test/unit/tasks/seeds_test.rb:18]: nice!

**Review by tbrisker — DISMISSED** [review]

File: test/test_helper.rb (hunk 5 above)
Comment (lzap) [reply test/test_helper.rb]: This was a copy paste from audited gem, sure I can do that. :-)

Comment (lzap) [reply test/test_helper.rb]: Oh the method is `enable_auditing` and not `enable_auditing=` actually.

```ruby
      def without_auditing
        auditing_was_enabled = auditing_enabled
        disable_auditing
        yield
      ensure
        enable_auditing if auditing_was_enabled
      end

      def disable_auditing
        self.auditing_enabled = false
      end

      def enable_auditing
        self.auditing_enabled = true
      end

```

File: test/unit/tasks/seeds_test.rb (hunk 6 above)
Comment (lzap) [reply test/unit/tasks/seeds_test.rb]: Good point, I haven't found any.

**Comment by lzap** [comment]
I believe no changes are required @tbrisker however tests were failing because my previous change, fixed that.

**Comment by lzap** [comment]
Ping anyone can do quick merge here? I believe this is RTM. @ares or @timogoebel

File: test/test_helper.rb (hunk 5 above)
Comment (timogoebel) [reply test/test_helper.rb]: I think @tbrisker has a valid point here, do we want to revert this to the original state?

Comment (lzap) [reply test/test_helper.rb]: Again, this method was taken from audited gem, there is no `enable_auditing=` method available. Please elaborate what's wrong.

Comment (timogoebel) [reply test/test_helper.rb]: I am wondering if `klass.disable_auditing unless auditing_was_enabled` is missing.

**Comment by lzap** [comment]
Removed `assert_not` change and `.gitignore` change.

File: test/test_helper.rb (hunk 5 above)
Comment (ares) [reply test/test_helper.rb]: I think this is still not resolved, original value is not being restored in ensure, I think this would be necessary `"auditing_was_enabled ? klass.enable_auditing : klass.disable_auditing"` in ensure block

perhaps link audited code you copied here?

Comment (lzap) [reply test/test_helper.rb]: NOW I get it! This code was from audited gem, but they had the opposite `with_auditing". Looks like I only need this then: `klass.disable_auditing unless auditing_was_enabled`

**Comment by lzap** [comment]
How about this?

**Review by ares — APPROVED** [review]
Tests failres are unrelated, all comments were IMHO addressed, looks good to me too

**Comment by ares** [comment]
Thanks @lzap, merging!

**Comment by timogoebel** [comment]
... what about @tbrisker and my comment about resetting the audit state?
I believe this was still pending here.

**Comment by ares** [comment]
@timogoebel I believe it has been resolved, it was changed to `disable unless was_enabled` logic, which is I believe correct. If we enable it, we ensure it's disabled back unless it wasn't enabled alreasy

**Comment by timogoebel** [comment]
@ares: Oh, I see it now. Sorry for the confusion. All good. Would have merged it earlier, but needed this explanation. 👼

**Comment by ares** [comment]
No worries, we talked with lzap this morning on IRV and the final version was pushed just few hours ago :-) thanks for keeping eyes on what goes in!

**Comment by lzap** [comment]
Yeah this kind of setter/getter is confusing. But it works.
